cJSON.c

```diff
@@ -564,14 +564,20 @@ static cJSON_bool print_number(const cJSON * const item, printbuffer * const out
     }
     else
     {
-        /* Try 15 decimal places of precision to avoid nonsignificant nonzero digits */
-        length = sprintf((char*)number_buffer, "%1.15g", d);
-
-        /* Check whether the original double can be recovered */
-        if ((sscanf((char*)number_buffer, "%lg", &test) != 1) || !compare_double((double)test, d))
-        {
-            /* If not, print with 17 decimal places of precision */
-            length = sprintf((char*)number_buffer, "%1.17g", d);
+        /* If decimal places are present and value is not an integer type */
+        if (item->decimal_places && (item->decimal_places < 15) && (ceil(d) != d)) {
```

Comment:
I don't think 0 should be the sentinel value, it's a valid possibility to want 0 decimal places (i.e. truncate to integer).  
Probably best to use an optional, an other bool field, or something like INT_MAX.

Reply:
Thanks for the review. I have different opinion on truncation of floating number. Simple truncation can be done by type casting of float to int. sprintf usually do rounding of float when fixed decimal places are requested which is not exactly a truncation. This is trivial but for the purpose I think its best to keep it as 0 for no decimal place request during printing.

Reply:
You are right regarding truncation/rounding with `sprintf`, thanks.

In https://github.com/DaveGamble/cJSON/issues/175#issuecomment-304819184 you wanted to use -1, which seems totally reasonable as well.  
I am not sure I understand your motivations for not allowing 0. It seems like convenience at no cost?

Reply:
Well, I think I wanted to keep things simple since cJSON_New_Item create and memset the cJSON object. one does not have to bother initializing decimal_place specially to make it a valid value. Hence I decided to keep it 0 as default and no preference.

If somebody is creating a static cJSON object then they have to make sure variables are initialized properly else this may create unknown issues.

What do you say?

Reply:
That's a fair point. I am personally usually angling towards having an external interface that makes sense, then deal with implementation details, but in this case it looks like it would break existing code.  
I guess it's up to the maintainer.

Reply:
The problem with having 0 as fixed places seems pointless since the behavior may become unpredictable to some users. Unless user know and want to round of a floating value to nearest integer, 0 has no meaning. Its not going to truncate as they may think and will become an error case for them.
So I do not see any issue to allow user to only provide a number between 1 to 14. 15+ being then default behavior.